File: Regression/Decision_Tree_Regression/Decision_Tree.py
"""
Program     : Decision Tree Regression
Date        : 5 Agustus 2018
Author      : Bayu Aditya
"""

# Import Data
import pandas as pd
database = pd.read_csv('Position_Salaries.csv')
level = database.iloc[:, 1:2].values                    # independent variable
gaji = database.iloc[:, 2:].values                      # dependent variable

# Regresi Decision Tree
from sklearn.tree import DecisionTreeRegressor
tree_reg = DecisionTreeRegressor().fit(level, gaji)     # Class Regresi Tree
y_pred = tree_reg.predict(level)                        # Regresi berdasarkan 'level'

# Visualisasi
import matplotlib.pyplot as plt
plt.scatter(level, gaji, color = 'red')
plt.plot(level, y_pred, color = 'blue')
plt.title('Decision Tree') ; plt.grid(True)
plt.xlabel('Level') ; plt.ylabel('Gaji')
plt.show()

# Regresi Decision Tree (HIGH RESOLUTION) n data
n = 100                                                 # n data
import numpy as np 
x_high = np.linspace(min(level), max(level), n)         # x_high
x_high = np.reshape(x_high, (n,1))                      # reshape matriks x_high
y_pred_high = tree_reg.predict(x_high)                  # Regresi berdasarkan 'x_high'

# Visualisasi (HIGH RESOLUTION)
plt.scatter(level, gaji, color = 'red')
plt.plot(x_high, y_pred_high, color = 'blue')
plt.title('Decision Tree (HIGH RESOLUTION)') ; plt.grid(True)
plt.xlabel('Level') ; plt.ylabel('Gaji')
plt.show()

File: Regression/Decision_Tree_Regression/Decision_Tree.R
'
Program     : Decision Tree Regression
Date        : 30 Jul 2018
Author      : Bayu Aditya
'

# Import data
database = read.csv('Position_Salaries.csv')
database = database[2:3]

# Regresi Decision Tree
# install.packages('rpart')
library(rpart)
dec_tree = rpart(formula = Salary ~. , 
                 data = database,
                 control = rpart.control(minsplit = 1))
# summary(dec_tree)
y_pred = predict(dec_tree, newdata = database)

# Plot Data dan hasil regresi
library(ggplot2)
ggplot() + 
  geom_point(aes(x = database$Level, y = database$Salary), color = 'red') +
  geom_line(aes(x = database$Level, y = y_pred), color = 'blue') +
  ggtitle('Decision Tree') +
  xlab('Level') +
  ylab('Salary')

# Visualising the Regression Model results (for higher resolution and smoother curve)
# install.packages('ggplot2')
library(ggplot2)
x_grid = seq(min(database$Level), max(database$Level), 0.1)
ggplot() +
  geom_point(aes(x = database$Level, y = database$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(dec_tree, newdata = data.frame(Level = x_grid))),
            colour = 'blue') +
  ggtitle('Decision Tree High Resolution') +
  xlab('Level') +
  ylab('Salary')

# Prediksi Sembarang nilai
y_coba = predict(dec_tree, data.frame(Level = 5.5))
